apollo-ast/src/main/kotlin/com/apollographql/apollo3/ast/internal/SchemaValidationScope.kt

```diff
@@ -185,3 +189,25 @@ private fun SchemaValidationScope.validateNoIntrospectionNames() {
   }
 }
 
+/**
+ * To prevent surprising behaviour, objects that declare key fields that also implement interfaces that declare key fields are an error
+ *
+ * @see <a href="https://github.com/apollographql/apollo-kotlin/issues/3356#issuecomment-1134381986">Discussion</a>
+ */
+private fun SchemaValidationScope.validateKeyFields(mergedDefinitions: List<GQLDefinition>) {
+  val mergedTypeDefinitions = mergedDefinitions.filterIsInstance<GQLTypeDefinition>().associateBy { it.name }
+  val keyFields = mergedTypeDefinitions.filterValues { it.canHaveKeyFields() }
+      .values.associate { it.name to it.keyFields(mergedTypeDefinitions) }
+
+  mergedTypeDefinitions.values.filterIsInstance<GQLObjectTypeDefinition>().forEach { o ->
+    if (keyFields[o.name]!!.isNotEmpty()) {
+      val interfacesWithKfs = o.implementsInterfaces.map { it to keyFields[it]!! }.filter { it.second.isNotEmpty() }.map { it.first }
+      if (interfacesWithKfs.isNotEmpty()) {
+        registerIssue(
+            "Type '${o.name}' cannot have key fields since it implements the following interfaces which also have key fields: $interfacesWithKfs",
+            sourceLocation = o.sourceLocation
```

Comment:
👍 my initial thoughts were "ensure the key fields are the same" but this is better.